### PR TITLE
BF: GitRepo.commit to work correctly on directory paths (Fixes #3087)

### LINKED ISSUE #3087: save fails when passed a directory
`save` doesn't work when given a directory as of 5ae188c89 (BF: possibly make use of temporary index every time (not only direct mode) committing a list of files, 2018-11-23, gh-3009).

```
% mkdir sdir && echo foo >sdir/foo

% datalad save sdir
CommandError: command '['git', '-c', 'receive.autogc=0', '-c', 'gc.auto=0', 'commit', '-m', '[DATALAD] Recorded existing changes']' failed with exitcode 1
Failed to run ['git', '-c', 'receive.autogc=0', '-c', 'gc.auto=0', 'commit', '-m', '[DATALAD] Recorded existing changes'] under '/tmp/tmp.DatFpoHPdN-dl'. Exit code=1. out=On branch master
Untracked files:
        sdir/

nothing added to commit but untracked files present
 err=
On branch master
Untracked files:
        sdir/

nothing added to commit but untracked files present

% git status
On branch master
Changes to be committed:
        new file:   sdir/foo
```

<details>

<summary>bisect script</summary>

```sh
#!/bin/sh

set -eux

tmpdir=$(mktemp -d --suffix=-savefail)
datalad create $tmpdir
cd $tmpdir

mkdir -p s
echo foo >s/a
datalad save s

test -z "$(git diff-index --cached HEAD -- s/a)" || exit 1
```

</summary>

### PR DESCRIPTION
**Pull request by yarikoptic**
This pull request fixes #3087 through comparing to the `git diff` for the files provided to the commit call, so should behave properly in case of directories etc.

### Changes
- [x] Needs fixing since breaks the tests
- Make checks optional for the `rev-` functionality since it cares not about staged or not and about direct mode, so I guess a plain `commit(files)` should suffice in that case - probably should better be done in master upon merge

### CODE REVIEW

**Comment by mih** [comment]
Please make the new behavior optional and (ideally) off by default. The -revolution code stack performs this kind of normalization very early on (it is needed in other places too). Doing it unconditionally this late in the game would have no benefit and slow things down in its context.

If this is not desirable, I can overwrite `commit()` in `RevolutionGitRepo`, but I wanted to mention it.

It might be worth investigating whether the `ls-files` call done here should be modified to honor `.gitignore` setttings too. You could have a look at the specific call done in `RevolutionGitRepo.get_content_info()` -- it took a few cycles to arrive at this particular setup.

**Comment by yarikoptic** [comment]
Thanks for the info/pointers!
Unless someone gets time before then, I really hope to push on #2926 in a week or two so may be this one would even be superseded ;)

File: datalad/support/gitrepo.py [hunk 1]
```diff
@@ -2624,8 +2609,23 @@ def get_changed_files(self, staged=False, diff_filter='', index_file=None):
             opts.append('--diff-filter=%s' % diff_filter)
         if index_file:
             kwargs['env'] = {'GIT_INDEX_FILE': index_file}
-        return [normpath(f)  # Call normpath to convert separators on Windows.
-                for f in self.repo.git.diff(*opts, **kwargs).split('\0') if f]
+        if files:
```
Comment (yarikoptic) [inline datalad/support/gitrepo.py]: surprised that test didn't fail (locally) -- I intended to make it `if files is not None` here and I thought I added explicit test (may be died in intermediate RF), TODO

**Comment by yarikoptic** [comment]
ok, @kyleam - should be ready for review.
I've done some limited benchmarking -- no notable change in comparison to 0.11.x ;)

File: datalad/support/gitrepo.py (hunk 1 above)
Comment (yarikoptic) [reply datalad/support/gitrepo.py]: done now in db0a844c691d6848ca9627e567e658e9f3722594

**Pull request merged**